packages/toc-extension/src/index.ts

```diff
@@ -42,212 +52,194 @@ namespace CommandIDs {
+        editorTracker,
+        toc,
+        rendermime.sanitizer,
+        translator,
+        settings
+      );
+      registry.add(markdownGenerator);
 
-  // Create a rendered Markdown generator:
-  if (markdownViewerTracker) {
-    const renderedMarkdownGenerator = createRenderedMarkdownGenerator(
-      markdownViewerTracker,
-      toc,
-      rendermime.sanitizer,
-      translator,
-      settings
-    );
-    registry.add(renderedMarkdownGenerator);
-  }
+      // Create a LaTeX generator:
+      const latexGenerator = createLatexGenerator(editorTracker);
+      registry.add(latexGenerator);
 
-  // Update the ToC when the active widget changes:
-  if (labShell) {
-    labShell.currentChanged.connect(onConnect);
-  }
+      // Create a Python generator:
+      const pythonGenerator = createPythonGenerator(editorTracker);
+      registry.add(pythonGenerator);
+    }
 
-  return registry;
-
-  /**
-   * Callback invoked when the active widget changes.
-   *
-   * @private
-   */
-  function onConnect() {
-    let widget = app.shell.currentWidget;
-    if (!widget) {
-      return;
+    // Create a rendered Markdown generator:
+    if (markdownViewerTracker) {
+      const renderedMarkdownGenerator = createRenderedMarkdownGenerator(
+        markdownViewerTracker,
+        toc,
+        rendermime.sanitizer,
+        translator,
+        settings
+      );
+      registry.add(renderedMarkdownGenerator);
     }
-    let generator = registry.find(widget);
-    if (!generator) {
-      // If the previously used widget is still available, stick with it.
-      // Otherwise, set the current ToC widget to null.
-      if (toc.current && toc.current.widget.isDisposed) {
-        toc.current = null;
+
+    const updateCurrent = () => {
+      const widget = app.shell.currentWidget;
+      if (!widget) {
+        return;
       }
-      return;
+      const generator = registry.find(widget);
+      if (!generator) {
+        // If the previously used widget is still available, stick with it.
+        // Otherwise, set the current ToC widget to null.
+        if (toc.current && toc.current.widget.isDisposed) {
+          toc.current = null;
+        }
+        return;
+      }
+      toc.current = { widget, generator };
+    };
+
+    // Update the ToC when the active widget changes
+    if (labShell) {
+      labShell.currentChanged.connect(updateCurrent);
     }
-    toc.current = { widget, generator };
-  }
-}
 
-/**
- * Initialization data for the ToC extension.
- *
- * @private
- */
-const extension: JupyterFrontEndPlugin<ITableOfContentsRegistry> = {
-  id: '@jupyterlab/toc:plugin',
-  autoStart: true,
-  provides: ITableOfContentsRegistry,
-  requires: [IDocumentManager, IRenderMimeRegistry, ITranslator],
-  optional: [
-    IEditorTracker,
-    ILayoutRestorer,
-    ILabShell,
-    IMarkdownViewerTracker,
-    INotebookTracker,
-    ISettingRegistry
-  ],
-  activate: activateTOC
+    app.started.then(updateCurrent);
```

Comment:
Need to check the order of events (widget added to the main area, app started) is as expected, so this line is relevant.

Reply:
This might work in the case of RetroLab which should add the widget on startup to the main area via the document opener.

However this might not be robust enough in more complex scenarios where widgets are added a later time. Maybe this toc plugin should then be reworked to allow for more flexibility like the file browser widget.

Reply:
Checking in RetroLab with a local build of https://github.com/jupyterlab/retrolab/pull/275:

![image](https://user-images.githubusercontent.com/591645/143595430-3093dfe9-3bba-4ea6-90ad-45ae5736f8cd.png)

This change is unfortunately not enough, and being able to listen to `currentChanged` seems to be required.

One way forward would be to decouple the widget and registry to allow for better reuse of the plugin(s).

Reply:
While at it - I was thinking whether LSP should be able populate the TOC widget with document outline in script files, or whether it should be a separate widget. I wonder if the decoupling you mention could help that use case.

Reply:
Probably yes. For LSP that could mean adding a new generator similar to the existing ones?

https://github.com/jupyterlab/jupyterlab/blob/51d8f101f7ab5df7551dcd32ec4cb47c10f2fb4d/packages/toc-extension/src/index.ts#L155-L197